### ISSUE TITLE
Language service

### ISSUE THREAD

**Issue by minestarks** [microsoft/qsharp#396]
This issue tracks the implementation of the language service and editor features for Q#. Editor features are things like autocomplete, go-to-definition and hover. The language service is the component that uses the compiler internals to keep track of the current state of the program, update the program in response to document updates, and answer questions about it in response to requests from the editor.

## Components involved

1. **Language service**: this is a Rust crate that contains the actual "smarts" of the language service, including the implementation of each individual feature such as completion, hover, etc. 
2. **JS bindings**: this is a trivial WASM wrapper over the language service to expose it as a JavaScript class. It follows a similar pattern to the existing WASM wrapper for the interpreter (the `run()` method etc)
3. **npm package**: This is the same package used by the playground that incorporates the Q# compiler. A new interface will be added to provide access to the language service methods, with the option to host the language service in a separate web worker (just like the interpreter today).
4. **VS Code extension**: Implementation of the VS Code extensibility APIs. These APIs generally map directly to language service methods.
5. **Playground**: Similar to VS Code, the language service methods just need to be wired up to the Monaco editor extension points in the playground.
5. **Python bindings**: Very similar to the WASM/JS bindings mentioned above - a trivial wrapper over the language server crate to expose it as a Python class. Also similar to the Python interpreter wrapper that currently exists today in the pip package.
7. **Python LSP server**: (for JupyterLab support) A standalone LSP ([Language Server Protocol](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#languageServerProtocol)) server implemented in Python. While Python may not be the most obvious choice for an LSP server implementation (Node.js libraries maintained by VS Code would have been a more natural fit), it was chosen to make the installation simple (user can just `pip install` the extension instead of having to separately install Node.js) for a JupyterLab user.
8. **Jupyterlab extension**: (for JupyterLab support) Includes the LSP server module and registers it with JupyterLab.

```mermaid
%%{init: { "flowchart": { "nodeSpacing": 3 } } }%%
flowchart TD

subgraph JS
  playground[Playground]
  vscode[VS Code extension]
  website[Website]
  npm[npm package]
end

subgraph Python
  jupyterlab[JupyterLab extension]
  lsp[LSP server]
end

subgraph Rust
  wasm["JS bindings (WASM)"]
  pip[Python bindings]
  interpreter[Interpreter]
  ls[Language service]
  compiler[Compiler]
end

playground
vscode
website
jupyterlab

playground-->npm
vscode-->npm
website-->npm
jupyterlab-->lsp

npm-->wasm
lsp-->pip
wasm-->interpreter
wasm-->ls
pip-->ls

interpreter-->compiler
ls-->compiler
```

## Implementation plan

This list shows the proposed order of implementation, based on complexity (simpler features like hover are prioritized above more complicated ones like completions) and feature priority (VS Code extension being the highest priority deliverable).

The links are to LSP documentation. Even if we don't implement a true LSP server, we'll follow the LSP spec while designing the interface as it closely resembles the extensibility model used by many editors (VS Code, Monaco, JupyterLab, etc).

- Scaffolding:
    - [x] Language service crate with minimal completion, hover, go-to-def and diagnostics (#371)
    - [x] npm API and WASM layer (#401) (#426)
    - [x] Wire up feature providers in VS Code and playground (#429)
- P1 features. These will be implemented with support for only the file currently opened in the editor. The whole Q# program gets recompiled with every document edit (i.e. no performance optimizations):
    - [x] [**Go to definition**](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_definition): 
        - [x] locations within the current file. (#460)
        - [x] For things defined in the standard library, create a virtual document showing the standard library source.
    - [x] [**Hover**](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_hover): Hover info for callables, newtypes, patterns at the declaration site as well as the usage site showing name, type and documentation.
        - [x] _Compiler dependency_: Parsing doc comments. (#386)
        - [x] Hover for UDT definitions and References. (#415, #519, #464, #449)
    - [x] [**Diagnostics**](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_publishDiagnostics): Error checking is currently implemented in the playground. Expose in VS Code by publishing diagnostics upon every recompilation (i.e. every keystroke in practice, barring any perf optimizations). (#394 (preliminary))
    - [x] [**Completions (basic)**](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_completion): Include all known symbols in the completion list. This is a "good enough" temporary implementation until we have improved error recovery in the parser and the ability to provide more accurate completions for each context. (#443)
    - [x] _Compiler dependency_: Error recovery in the parser - ability to get information about an imperfect program. (#380)
    - [x] **Completions (better)**: More sophisticated completion list behavior. We could leverage the parser here to narrow down the next expected syntax kind (e.g. type, expression) and populate the completion list based on the tokens and known names (e.g. callables, locals) that are applicable to that syntax.
      - [x] _Compiler dependency_: parser-provided completions
      - [x] [Completion item details](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem_resolve) (type, documentation). (#511)
    - [x] [**Signature help**](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_signatureHelp): Signature help for function calls including parameter documentation.
- Notebook cell support. Notebook cells use a variant of Q# that allows top-level items and expressions. 
  - [x] Use the "fragment" compiler rather than the "regular" compiler to construct a program.
  - [x] There may be a _compiler dependency_ here (investigate)
- P2 features. Some of these features bring new complexity as they can return [text edits](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textEdit) back to the client.
    - [x] [**Format document**](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_formatting)
    - [x] **[Format range](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_rangeFormatting), [format on type](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_onTypeFormatting)** - Ability to format the subset of a document, as well the applicable range in response to a trigger character (e.g. `}`)
    - [X] [**Find all references**](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_references) - essentially reverse go-to-definition. (#830 )
    - [x] [**Symbol rename**](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_rename) - determine if a rename is valid in the current location, use find-all-references to locate references to symbol, return text edits for rename
    - [x] [**Quick fixes**](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_codeAction) Contrast with _refactorings_, which are designated as a P3 feature below. Code fixes are associated with diagnostics (e.g. "unused identifier") reported by the compiler/linter and provide text edits that would "fix" that diagnostic.
      - [x] _Compiler dependency_: Lint-level diagnostics for desired quick fixes.
- Support for closed files (only necessary when we support file references (modules, imports etc) and/or projects):
    - [x] VS Code Web compatible File I/O using vscode's WASI implementation, mapping of utf-16 line/column offsets
    - [ ] File watching - updating compilation in response to workspace changes (e.g. git branch switch)
- LSP server to enable support for other editors
    - [ ] Python LSP server and JupyterLab support (blocked on [jupyterlab-lsp](https://jupyterlab-lsp.readthedocs.io/en/latest/Language%20Servers.html) supporting the latest version of JupyterLab).
- Performance optimizations, as appropriate. These come with tradeoffs, so we should only implement them if we're observing actual performance problems. 
    - [ ] Debouncing/batching - strategies to reduce the number of recompilations by introducing a slight delay before recompiling the document on each keystroke/file update. This may introduce a slight delay to completions and error squiggles. 
    - [ ] Cancellation - cancel an outstanding recompilation when a document is updated.  May require web workers or some other threading mechanism.
    - [ ] Background checking - ability for language service requests to run concurrently with checking. Requires web workers for WASM.
- P3 features:
    - [ ] [**Go to symbol**](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_symbol)
    - [ ] [**Refactorings**](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_codeAction) - e.g. extract function. Contrast with quick fixes described above.
    - [ ] [**Semantic highlighting**](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens) - e.g. highlighting of resolved identifiers based on type
    - [ ] [**Document highlights**](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_documentHighlight)
    - [ ] ...and so on.
- Even more expensive performance optimizations (hopefully avoided):
    - [ ] Background program update - ability to work with an immutable snapshot of the program as updates are being made in the background (in response to keystrokes or file updates). Requires web workers for WASM.
    - [ ] _Compiler dependency_: Partial recompilation - ability to recompile only parts of a program as the document gets updated.

**Comment by minestarks** [comment]
> @minestarks -- is #1038 going to unblock LSP support, or are there other changes required?

@sezna well, we'd have to implement an LSP server... but if you're asking if anything in the core language service crate needs to change, I don't think so. I had a prototype for this a few months ago using the https://github.com/python-lsp/python-lsp-jsonrpc package which was working in JupyterLab. I think it's just a matter of wiring pieces together.

> Also, perhaps we could file some separate issues for some of these features and use this issue to track them?

Yes this issue is quite big, I don't mind breaking off a few items. I didn't want to create issues for things that aren't even on our radar, but for near-term stuff (e.g. formatting) and for stuff we want a dedicated discussion thread for, it makes sense.

**Comment by minestarks** [comment]
Closing as completed. The unchecked items in the original description are currently out of scope.
